File: lib/tools/caption.ts
import Anthropic from "@anthropic-ai/sdk";
import { env } from "../config";
import { CAPTION_GENERATION_PROMPT } from "../prompts";

let _client: Anthropic | null = null;
function client(): Anthropic {
  if (!_client) _client = new Anthropic({ apiKey: env.ANTHROPIC_API_KEY });
  return _client;
}

export async function generateCaption(
  subject: string,
  context = "general tribute",
): Promise<{ caption: string; hashtags: string }> {
  const response = await client().messages.create({
    model: env.ANTHROPIC_MODEL,
    max_tokens: 500,
    messages: [{ role: "user", content: CAPTION_GENERATION_PROMPT(subject, context) }],
  });

  const text =
    response.content
      .map((b) => (b.type === "text" ? b.text : ""))
      .join("")
      .trim();

  const cleaned = stripFences(text);
  try {
    const parsed = JSON.parse(cleaned) as { caption?: string; hashtags?: string };
    if (parsed.caption && parsed.hashtags) {
      return { caption: parsed.caption.trim(), hashtags: parsed.hashtags.trim() };
    }
  } catch {
    /* fall through */
  }

  return {
    caption: `Thank you ${subject}.`,
    hashtags: "#bandtoons #cartoon #vintageart #rubberhose #music",
  };
}

function stripFences(text: string): string {
  let t = text.trim();
  if (t.startsWith("```")) {
    t = t.slice(3);
    if (t.startsWith("json")) t = t.slice(4);
    const end = t.lastIndexOf("```");
    if (end !== -1) t = t.slice(0, end);
  }
  return t.trim();
}

File: src/tools/caption.py
import json
import logging

from anthropic import AsyncAnthropic

from ..config import settings
from ..prompts import CAPTION_GENERATION_PROMPT

logger = logging.getLogger(__name__)

_client = AsyncAnthropic(api_key=settings.anthropic_api_key)


async def generate_caption(subject: str, context: str = "tribute / homage") -> dict:
    """Generate caption + hashtags via Claude. Returns dict with `caption` and `hashtags`."""
    prompt = CAPTION_GENERATION_PROMPT.format(subject=subject, context=context)

    response = await _client.messages.create(
        model=settings.anthropic_model,
        max_tokens=500,
        messages=[{"role": "user", "content": prompt}],
    )

    text = response.content[0].text.strip()
    # Strip code fences if present
    if text.startswith("```"):
        text = text.split("```")[1]
        if text.startswith("json"):
            text = text[4:]
    text = text.strip()

    try:
        data = json.loads(text)
        return {
            "caption": data["caption"].strip(),
            "hashtags": data["hashtags"].strip(),
        }
    except (json.JSONDecodeError, KeyError) as e:
        logger.error("Failed to parse caption JSON: %s | raw: %s", e, text)
        # Fallback
        return {
            "caption": f"Thank you {subject}.",
            "hashtags": "#bandtoons #cartoon #vintageart #rubberhose #music",
        }
